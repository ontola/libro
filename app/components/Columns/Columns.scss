@import '../shared/config';

.Columns {
  align-items: flex-start;
  box-sizing: border-box;

  @include large {
    display: flex;
    margin-left: -1rem;
    margin-right: -1rem;

    &--flexGrow > .Column {
      flex-grow: 1;
      margin: units('none', 'large');
      width: 100%;
    }
  }
}

.Column {
  box-sizing: border-box;
}

@each $size, $margin in $sizes-margin {
  .Columns--gutter-#{$size} {
    margin-bottom: 1rem;
    margin-top: 1rem;

    @include large {
      margin-left: -$margin;
      margin-right: -$margin;

      & > .Column {
        margin-left: $margin;
        margin-right: $margin;
      }
    }
  }
}
