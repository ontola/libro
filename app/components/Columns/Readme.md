The Column component takes all direct children and makes columns out of them.

Make sure to give each child a `key` property to ensure proper rendering.

      <Columns>
        <Card key="col-1">Column 1</Card>
        <Card key="col-1">Column 2</Card>
      </Columns>

Custom flexBasis (for narrower columns)

      <Columns flexBasis="0">
        <Card key="col-1">Column 1</Card>
        <Card key="col-2">Column 2</Card>
        <Card key="col-3">Column 3</Card>
        <Card key="col-4">Column 4</Card>
      </Columns>
