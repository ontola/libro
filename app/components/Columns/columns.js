import './columns.scss';
import React from 'react';

export default class Columns extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div className="cols">{this.props.children}</div>
    );
  }
}

Columns.propTypes = {
  children: React.PropTypes.node.isRequired
};
