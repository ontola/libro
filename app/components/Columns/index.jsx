import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

import { sizes } from '../shared/config';

import './Columns.scss';

const propTypes = {
  /** Each child becomes a column. */
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.node,
  ]).isRequired,
  flexBasis: PropTypes.string,
  flexGrow: PropTypes.bool,
  gutter: PropTypes.oneOf(sizes),
  size: PropTypes.oneOf(sizes),
};

const defaultProps = {
  flexBasis: '19em',
  flexGrow: true,
  gutter: 'medium',
};

const Columns = ({
  children,
  flexBasis,
  flexGrow,
  gutter,
  size,
}) => {
  const className = classNames({
    Columns,
    'Columns--flex-grow': flexGrow,
    [`Columns--gutter-${gutter}`]: gutter,
    [`Columns--size-${size}`]: size,
  });

  const normChildren = Array.isArray(children) ? children : [children];
  const renderColumns = normChildren.map(column => (
    <div
      className="Column"
      key={`col-${column.key}`}
      style={{
        flexBasis,
      }}
    >
      {column}
    </div>
  ));

  return (
    <div className={className}>{renderColumns}</div>
  );
};

Columns.propTypes = propTypes;
Columns.defaultProps = defaultProps;

export default Columns;
