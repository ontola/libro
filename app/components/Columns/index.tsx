import React from 'react';
import clsx from 'clsx';

import Column from '../Column';
import { Size } from '../shared/config';

import './Columns.scss';

interface ColumnsProps {
  /** Each child becomes a column. */
  children: JSX.Element | JSX.Element[];
  flexBasis?: string;
  flexGrow?: boolean;
  gutter?: Size;
  size?: Size;
}

const defaultProps = {
  flexBasis: '19em',
  flexGrow: true,
  gutter: 'small',
};

const Columns = ({
  children,
  flexBasis,
  flexGrow,
  gutter,
  size,
}: ColumnsProps): JSX.Element => {
  const className = clsx({
    Columns,
    'Columns--flex-grow': flexGrow,
    [`Columns--gutter-${gutter}`]: gutter,
    [`Columns--size-${size}`]: size,
  });

  const normChildren = Array.isArray(children) ? children : [children];
  const renderColumns = normChildren.map((child: JSX.Element) => (
    <Column
      flexBasis={flexBasis}
      key={`col-${child.key}`}
    >
      {child}
    </Column>
  ));

  return (
    <div className={className}>
      {renderColumns}
    </div>
  );
};

Columns.defaultProps = defaultProps;

export default Columns;
