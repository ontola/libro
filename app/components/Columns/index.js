import './columns.scss';
import React, { PropTypes } from 'react';

function Columns({ children }) {
  return (
    <div className="cols">{children}</div>
  );
}

Columns.propTypes = {
  children: PropTypes.node.isRequired,
};

Columns.defaultProps = {
  children: [],
};

export default Columns;
