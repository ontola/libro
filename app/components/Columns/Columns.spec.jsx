import React from 'react';

import Columns from './';

const exampleChildren = [<div>Joe</div>, <div>Joe</div>];

argUnit(Columns, () => {
  describe('with node child', () => {
    set('children', () => <p>test</p>);

    it('renders', () => {
      expect(subject).toMatchSnapshot();
    });
  });

  describe('with array child', () => {
    set('children', () => exampleChildren);

    it('renders', () => {
      expect(subject).toMatchSnapshot();
    });

    it('renders all children', () => {
      expect(subject.children()).toHaveLength(exampleChildren.length);
    });
  });
});
