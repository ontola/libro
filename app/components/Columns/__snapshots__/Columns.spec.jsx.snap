// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns with array child renders 1`] = `
<div
  className="Columns Columns--flex-grow Columns--gutter-medium"
>
  <div
    className="Column"
    key="col-null"
    style={
      Object {
        "flexBasis": "19em",
      }
    }
  >
    <div>
      Joe
    </div>
  </div>
  <div
    className="Column"
    key="col-null"
    style={
      Object {
        "flexBasis": "19em",
      }
    }
  >
    <div>
      Joe
    </div>
  </div>
</div>
`;

exports[`Columns with node child renders 1`] = `
<div
  className="Columns Columns--flex-grow Columns--gutter-medium"
>
  <div
    className="Column"
    key="col-null"
    style={
      Object {
        "flexBasis": "19em",
      }
    }
  >
    <p>
      test
    </p>
  </div>
</div>
`;
